### PR TITLE
fixes #22664; guard against potential seqs self assignments

### LINKED ISSUE #22664: =copy hook for seq appears to be broken
### Description

When wrapping MAPR I wanted to test my hooks with Valgrind to make sure I had gotten them right. What I ended up with was finding a bug in the `=copy` hook for sequences placed in ref objects. The smallest reproduction I've been able to create is this:

```nim
type
  ElementKind* = enum String, Number
  Element* = object
    case kind*: ElementKind
    of String:
      str*: string
    of Number:
      num*: float
  Calc = ref object
    stack: seq[Element]

var calc = new Calc

calc.stack.add Element(kind: Number, num: 200.0)
echo "Calc stack before: ", calc.stack
echo "---"
let calc2 = calc
calc2.stack = calc.stack # This nulls out the object in the stack
echo "Calc stack after: ", calc.stack
echo "Calc2 stack after: ", calc2.stack
```

If a simple `seq[int]` is used the program behaves as expected. So it appears to only happen for more complex objects. The reason is that `=copy` is triggered on `calc.stack` to `calc2.stack` and the hook doesn't check if they are already equal.

### Nim Version

Both 2.0.0 and the current devel

### Current Output

```text
Calc stack before: @[(kind: Number, num: 200.0)]
---
Calc stack after: @[(kind: String, str: "")]
Calc2 stack after: @[(kind: String, str: "")]
```

### Expected Output

```text
Calc stack before: @[(kind: Number, num: 200.0)]
---
Calc stack after: @[(kind: Number, num: 200.0)]
Calc2 stack after: @[(kind: Number, num: 200.0)]
```

### Possible Solution

_No response_

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by ringabout**
fixes #22664

### CODE REVIEW

File: compiler/liftdestructors.nim [hunk 1]
```diff
@@ -553,17 +553,29 @@ proc forallElements(c: var TLiftCtx; t: PType; body, x, y: PNode) =
...
+            x,
+            y)
+  cond.typ = getSysType(c.g, c.info, tyBool)
+  body.add genIf(c, cond, newTreeI(nkReturnStmt, c.info, newNodeI(nkEmpty, c.info)))
+
 proc fillSeqOp(c: var TLiftCtx; t: PType; body, x, y: PNode) =
   case c.kind
   of attachedDup:
+    # similar as attachedAsgn
+    checkSelfAssignment(c, t, body, x, y)
```
Comment (Araq) [inline compiler/liftdestructors.nim]: Why? You don't have a `y` here.

File: compiler/ast.nim [hunk 2]
```diff
@@ -722,7 +722,7 @@ type
     mInstantiationInfo, mGetTypeInfo, mGetTypeInfoV2,
     mNimvm, mIntDefine, mStrDefine, mBoolDefine, mGenericDefine, mRunnableExamples,
     mException, mBuiltinType, mSymOwner, mUncheckedArray, mGetImplTransf,
-    mSymIsInstantiationOf, mNodeId, mPrivateAccess, mZeroDefault
+    mSymIsInstantiationOf, mNodeId, mPrivateAccess, mZeroDefault, mSameSeqPayload
```
Comment (Araq) [inline compiler/ast.nim:725]: Does not require a magic, a codegen proc is good enough.

File: compiler/liftdestructors.nim (hunk 1 above)
Comment (ringabout) [inline compiler/liftdestructors.nim]:
```suggestion
```

Comment (ringabout) [reply compiler/liftdestructors.nim]: My bad

**Comment by ringabout** [comment]
Succeeded by https://github.com/nim-lang/Nim/pull/22671

File: compiler/ast.nim (hunk 2 above)
Comment (ringabout) [reply compiler/ast.nim:725]: => https://github.com/nim-lang/Nim/pull/22671

**Pull request closed without merging**